[Task] Add namespace crud

sufuf3 opened the pull request:
I finished Namespace CRUD, please help me review this PR.
Thank you.

sufuf3 committed:
Add newNamespaceService in server/route

To support CRUD of namespace, update src/server/route.go & add
newNamespaceService.

sufuf3 committed:
Add namespace entity for server handler

Ref:
https://raw.githubusercontent.com/kubernetes/website/master/content/en/examples/admin/namespace-dev.json
```
{
  "kind": "Namespace",
  "apiVersion": "v1",
  "metadata": {
    "name": "development",
    "labels": {
      "name": "development"
    }
  }
}
```

sufuf3 committed:
Add src/kubernetes/namespaces for kubeCtl

sufuf3 committed:
Add src/namespace/ for server handler package

sufuf3 committed:
Add namespace server handler

John-Lin commented on src/entity/namespace.go:
```diff
@@ -0,0 +1,25 @@
+
+// the const for NamespaceCollectionName
+const (
+	NamespaceCollectionName string = "namespace"
```
`namespaces`

John-Lin commented on src/entity/namespace.go:
```diff
@@ -0,0 +1,25 @@
+type Namespace struct {
+	ID        bson.ObjectId     `bson:"_id,omitempty" json:"id" validate:"-"`
+	Name      string            `bson:"name" json:"name" validate:"required,k8sname"`
+	Labels    map[string]string `bson:"labels,omitempty" json:"labels" validate:"required,dive,keys,alphanum,endkeys,required,alphanum"`
```
I think we don't need labels here

John-Lin commented on src/kubernetes/namespaces.go:
```diff
@@ -0,0 +1,35 @@
+
+// DeleteNamespace will delete the namespace by the namespace name
+func (kc *KubeCtl) DeleteNamespace(name string) error {
+	deletePolicy := metav1.DeletePropagationBackground
```
Do we need DeletePropagationBackground option here?

sufuf3 replied:
I referred to https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/namespace/deletion/namespaced_resources_deleter.go#L340-L344
IMO, maybe it's good to need?
What do you think?

John-Lin replied:
I think we should use **Foreground** waiting for all resource to be killed. see https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#DeletionPropagation

```
The object exists in the key-value store until the garbage collector
deletes all the dependents whose ownerReference.blockOwnerDeletion=true
from the key-value store.  API sever will put the "foregroundDeletion"
finalizer on the object, and sets its deletionTimestamp.  This policy is
cascading, i.e., the dependents will be deleted with Foreground.
```

John-Lin commented on src/server/handler_namespace.go:
```diff
@@ -0,0 +1,175 @@
+
+	if err := namespace.CreateNamespace(sp, &n); err != nil {
+		if errors.IsAlreadyExists(err) {
+			response.Conflict(req.Request, resp.ResponseWriter, fmt.Errorf("Namespace Name: %s already existed", n.Name))
```
`Namespace: %s already existed`

John-Lin commented on src/server/handler_namespace.go:
```diff
@@ -0,0 +1,175 @@
+	}
+
+	if err := namespace.DeleteNamespace(sp, &n); err != nil {
+		if errors.IsAlreadyExists(err) {
```
```
if errors.IsNotFound(err) {
```

sufuf3 committed:
Remove labels of namespace

sufuf3 committed:
Update DeletionPropagation from background to foreground

Use Foreground waiting for all resource to be killed.
Ref:
- https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#controlling-how-the-garbage-collector-deletes-dependents
- https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#DeletionPropagation

sufuf3 committed:
Fix CI test fail

John-Lin commented:
@sufuf3 Please add namespace api usage in API.md

sufuf3 committed:
Add Namespace CRUD in API.md

sufuf3 commented:
@John-Lin I added namespace api usage in API.md and modified the files.
Please help me review this PR. Thank you.

John-Lin commented on API.md:
````diff
@@ -1361,3 +1366,94 @@ Response Data:
+}
+```
+
+Response Data:
````
response data should be the namespace entity not create success message

sufuf3 replied:
I copied from service, sorry that I forgot to change the response data.
Updated it.

WakeupTsai commented on src/namespace/namespace.go:
```diff
@@ -0,0 +1,25 @@
+	return err
+}
+
+// DeleteNamespace willl delete namespace
```
小錯字 will

John-Lin replied:
我修了

John-Lin committed:
Update namespace.go
